File: client-android/serenada-call-ui/src/main/java/app/serenada/callui/CallAudioRoutes.kt
package app.serenada.callui

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.VolumeUp
import androidx.compose.material.icons.filled.Bluetooth
import androidx.compose.material.icons.filled.DirectionsCar
import androidx.compose.material.icons.filled.Headset
import androidx.compose.material.icons.filled.PhoneInTalk
import androidx.compose.material.icons.filled.Usb
import androidx.compose.ui.graphics.vector.ImageVector
import app.serenada.core.call.AudioDevice
import app.serenada.core.call.AudioDeviceDirection
import app.serenada.core.call.AudioDeviceKind
import app.serenada.core.call.AudioDeviceStatus
import app.serenada.core.call.BluetoothProfile
import java.util.Locale

internal fun currentCallAudioRoute(
    currentAudioDevice: AudioDevice?,
    availableAudioDevices: List<AudioDevice>,
): AudioDevice? {
    val currentKey = currentAudioDevice?.takeIf { it.isCallAudioOutputRoute() }?.callAudioRouteKey()
    val options = callAudioRouteOptions(currentAudioDevice, availableAudioDevices)
    return options.preferredActiveCallAudioRoute()
        ?: currentKey?.let { key ->
            options.firstOrNull { it.callAudioRouteKey() == key }
        }
}

internal fun callAudioRouteOptions(
    currentAudioDevice: AudioDevice?,
    availableAudioDevices: List<AudioDevice>,
): List<AudioDevice> {
    val candidates = buildList {
        addAll(availableAudioDevices)
        currentAudioDevice?.let(::add)
    }
        .filter { it.isCallAudioOutputRoute() }
    val activeKeys = candidates
        .filter { it.status == AudioDeviceStatus.ACTIVE || it == currentAudioDevice }
        .map { it.callAudioRouteKey() }
        .toSet()
    val devicesByRoute = linkedMapOf<String, AudioDevice>()
    candidates.forEach { device ->
        val key = device.callAudioRouteKey()
        val existing = devicesByRoute[key]
        devicesByRoute[key] = if (existing == null) {
            device
        } else {
            existing.preferredCallAudioRouteDisplay(device)
        }
    }
    return devicesByRoute.values
        .map { device ->
            if (device.callAudioRouteKey() in activeKeys) {
                device.copy(status = AudioDeviceStatus.ACTIVE)
            } else {
                device
            }
        }
        .sortedWith(
            compareBy<AudioDevice> { it.callAudioRouteSortRank() }
                .thenBy { it.callAudioRouteSortLabel().lowercase(Locale.ROOT) }
        )
}

internal fun AudioDevice.callAudioRouteKey(): String {
    val routeName = displayName.trim()
    val fallback = id.ifBlank { routeName }
    return when (kind) {
        is AudioDeviceKind.Speakerphone -> "speakerphone"
        is AudioDeviceKind.Earpiece -> "earpiece"
        is AudioDeviceKind.Bluetooth -> "bluetooth:$fallback"
        is AudioDeviceKind.WiredHeadset -> "wired"
        is AudioDeviceKind.CarAudio -> "car:$fallback"
        is AudioDeviceKind.Usb -> "usb:$fallback"
        is AudioDeviceKind.Other -> "other:$fallback"
    }
}

internal fun AudioDevice.callAudioRouteLabel(strings: Map<SerenadaString, String>?): String {
    val routeName = displayName.trim().takeIf { it.isNotEmpty() }
    return when (kind) {
        is AudioDeviceKind.Speakerphone -> resolveString(SerenadaString.CallAudioSpeaker, strings)
        is AudioDeviceKind.Earpiece -> resolveString(SerenadaString.CallAudioPhone, strings)
        is AudioDeviceKind.WiredHeadset -> routeName ?: resolveString(SerenadaString.CallAudioHeadset, strings)
        is AudioDeviceKind.Bluetooth -> routeName ?: resolveString(SerenadaString.CallAudioBluetooth, strings)
        is AudioDeviceKind.CarAudio -> routeName ?: resolveString(SerenadaString.CallAudioCar, strings)
        is AudioDeviceKind.Usb -> routeName ?: resolveString(SerenadaString.CallAudioUsb, strings)
        is AudioDeviceKind.Other -> routeName
            ?: resolveString(SerenadaString.CallAudioUnknown, strings)
    }
}

internal fun callAudioRouteIcon(kind: AudioDeviceKind?): ImageVector {
    return when (kind) {
        is AudioDeviceKind.Speakerphone -> Icons.AutoMirrored.Filled.VolumeUp
        is AudioDeviceKind.Earpiece -> Icons.Default.PhoneInTalk
        is AudioDeviceKind.WiredHeadset -> Icons.Default.Headset
        is AudioDeviceKind.Bluetooth -> Icons.Default.Bluetooth
        is AudioDeviceKind.CarAudio -> Icons.Default.DirectionsCar
        is AudioDeviceKind.Usb -> Icons.Default.Usb
        is AudioDeviceKind.Other, null -> Icons.AutoMirrored.Filled.VolumeUp
    }
}

private fun AudioDevice.isCallAudioOutputRoute(): Boolean {
    return direction == AudioDeviceDirection.OUTPUT || direction == AudioDeviceDirection.BOTH
}

private fun List<AudioDevice>.preferredActiveCallAudioRoute(): AudioDevice? {
    return filter { it.status == AudioDeviceStatus.ACTIVE }
        .minWithOrNull(
            compareBy<AudioDevice> { it.callAudioRouteActiveRank() }
                .thenBy { it.callAudioRouteSortLabel().lowercase(Locale.ROOT) }
        )
}

private fun AudioDevice.callAudioRouteActiveRank(): Int {
    return when (kind) {
        is AudioDeviceKind.Bluetooth -> 0
        is AudioDeviceKind.WiredHeadset -> 1
        is AudioDeviceKind.CarAudio,
        is AudioDeviceKind.Usb -> 2
        is AudioDeviceKind.Speakerphone -> 3
        is AudioDeviceKind.Earpiece -> 4
        is AudioDeviceKind.Other -> 5
    }
}

private fun AudioDevice.callAudioRouteSortRank(): Int {
    return when (kind) {
        is AudioDeviceKind.Speakerphone -> 0
        is AudioDeviceKind.Earpiece -> 1
        is AudioDeviceKind.Bluetooth -> 2
        is AudioDeviceKind.WiredHeadset -> 3
        is AudioDeviceKind.CarAudio,
        is AudioDeviceKind.Usb,
        is AudioDeviceKind.Other -> 4
    }
}

private fun AudioDevice.callAudioRouteSortLabel(): String {
    return displayName.trim().ifBlank { id }
}

private fun AudioDevice.preferredCallAudioRouteDisplay(candidate: AudioDevice): AudioDevice {
    val existingRank = callAudioRouteDisplayRank()
    val candidateRank = candidate.callAudioRouteDisplayRank()
    return when {
        candidateRank < existingRank -> candidate
        candidateRank > existingRank -> this
        candidate.status == AudioDeviceStatus.ACTIVE && status != AudioDeviceStatus.ACTIVE -> candidate
        else -> this
    }
}

private fun AudioDevice.callAudioRouteDisplayRank(): Int {
    val hasName = displayName.trim().isNotEmpty()
    return when (val kind = kind) {
        is AudioDeviceKind.Bluetooth -> when (kind.profile) {
            BluetoothProfile.HFP,
            BluetoothProfile.BLE -> if (hasName) 0 else 3
            BluetoothProfile.UNKNOWN -> if (hasName) 1 else 4
            BluetoothProfile.A2DP -> if (hasName) 2 else 5
        }
        is AudioDeviceKind.Speakerphone,
        is AudioDeviceKind.Earpiece -> 0
        is AudioDeviceKind.WiredHeadset,
        is AudioDeviceKind.CarAudio,
        is AudioDeviceKind.Usb,
        is AudioDeviceKind.Other -> if (hasName) 0 else 1
    }
}

File: client-ios/SerenadaCallUI/Sources/CallAudioRoutes.swift
import Foundation
import SerenadaCore

func currentCallAudioRoute(
    currentAudioDevice: AudioDevice?,
    availableAudioDevices: [AudioDevice]
) -> AudioDevice? {
    let currentKey = currentAudioDevice.flatMap { device in
        device.isCallAudioOutputRoute ? callAudioRouteKey(device) : nil
    }
    let options = callAudioRouteOptions(
        currentAudioDevice: currentAudioDevice,
        availableAudioDevices: availableAudioDevices
    )
    if let activeRoute = preferredActiveCallAudioRoute(options) {
        return activeRoute
    }
    if let currentKey, let currentRoute = options.first(where: { callAudioRouteKey($0) == currentKey }) {
        return currentRoute
    }
    return nil
}

func callAudioRouteOptions(
    currentAudioDevice: AudioDevice?,
    availableAudioDevices: [AudioDevice]
) -> [AudioDevice] {
    let candidates = (availableAudioDevices + [currentAudioDevice].compactMap { $0 })
        .filter(\.isCallAudioOutputRoute)
    let visibleCandidates = candidates.contains(where: { device in
        if case .bluetooth(_) = device.kind {
            return device.status == .active || device.status == .connecting || device == currentAudioDevice
        }
        return false
    })
        ? candidates.filter { $0.kind != .earpiece }
        : candidates
    let activeKeys = Set(
        visibleCandidates
            .filter { $0.status == .active || $0 == currentAudioDevice }
            .map(callAudioRouteKey)
    )
    var devicesByRoute: [String: AudioDevice] = [:]
    for device in visibleCandidates {
        let key = callAudioRouteKey(device)
        if let existing = devicesByRoute[key] {
            devicesByRoute[key] = preferredCallAudioRouteDisplay(existing, candidate: device)
        } else {
            devicesByRoute[key] = device
        }
    }
    return devicesByRoute.values
        .map { device in
            activeKeys.contains(callAudioRouteKey(device))
                ? callAudioRouteWithStatus(device, .active)
                : device
        }
        .sorted { lhs, rhs in
            let lhsRank = callAudioRouteSortRank(lhs.kind)
            let rhsRank = callAudioRouteSortRank(rhs.kind)
            if lhsRank != rhsRank { return lhsRank < rhsRank }
            return callAudioRouteSortLabel(lhs).localizedCaseInsensitiveCompare(
                callAudioRouteSortLabel(rhs)
            ) == .orderedAscending
        }
}

func callAudioRouteLabel(_ device: AudioDevice, strings: [SerenadaString: String]?) -> String {
    let routeName = device.displayName.trimmingCharacters(in: .whitespacesAndNewlines)
    func routeNameOr(_ key: SerenadaString) -> String {
        routeName.isEmpty ? resolveString(key, overrides: strings) : routeName
    }

    switch device.kind {
    case .speakerphone:
        return resolveString(.callAudioSpeaker, overrides: strings)
    case .earpiece:
        return resolveString(.callAudioPhone, overrides: strings)
    case .wiredHeadset:
        return routeNameOr(.callAudioHeadset)
    case .bluetooth(_):
        return routeNameOr(.callAudioBluetooth)
    case .carAudio:
        return routeNameOr(.callAudioCar)
    case .usb:
        return routeNameOr(.callAudioUsb)
    case .other:
        return routeNameOr(.callAudioUnknown)
    }
}

func callAudioRouteSystemImage(_ kind: AudioDeviceKind?) -> String {
    switch kind {
    case .speakerphone:
        return "speaker.wave.2.fill"
    case .earpiece:
        return "phone.fill"
    case .wiredHeadset:
        return "headphones"
    case .bluetooth(_):
        return "dot.radiowaves.left.and.right"
    case .carAudio:
        return "car.fill"
    case .usb:
        return "cable.connector"
    case .other, nil:
        return "speaker.wave.2.fill"
    }
}

func callAudioRouteKey(_ device: AudioDevice) -> String {
    let routeName = device.displayName.trimmingCharacters(in: .whitespacesAndNewlines)
    let fallback = device.id.isEmpty ? routeName : device.id

    switch device.kind {
    case .speakerphone:
        return "speakerphone"
    case .earpiece:
        return "earpiece"
    case .wiredHeadset:
        return "wired"
    case .bluetooth(_):
        return "bluetooth:\(fallback)"
    case .carAudio:
        return "car:\(fallback)"
    case .usb:
        return "usb:\(fallback)"
    case .other:
        return "other:\(fallback)"
    }
}

private func callAudioRouteSortRank(_ kind: AudioDeviceKind) -> Int {
    switch kind {
    case .speakerphone:
        return 0
    case .earpiece:
        return 1
    case .bluetooth(_):
        return 2
    case .wiredHeadset:
        return 3
    case .carAudio, .usb, .other:
        return 4
    }
}

private func callAudioRouteSortLabel(_ device: AudioDevice) -> String {
    let routeName = device.displayName.trimmingCharacters(in: .whitespacesAndNewlines)
    return routeName.isEmpty ? device.id : routeName
}

private func preferredActiveCallAudioRoute(_ devices: [AudioDevice]) -> AudioDevice? {
    devices
        .filter { $0.status == .active }
        .min { lhs, rhs in
            let lhsRank = callAudioRouteActiveRank(lhs.kind)
            let rhsRank = callAudioRouteActiveRank(rhs.kind)
            if lhsRank != rhsRank { return lhsRank < rhsRank }
            return callAudioRouteSortLabel(lhs).localizedCaseInsensitiveCompare(
                callAudioRouteSortLabel(rhs)
            ) == .orderedAscending
        }
}

private func callAudioRouteActiveRank(_ kind: AudioDeviceKind) -> Int {
    switch kind {
    case .bluetooth(_):
        return 0
    case .wiredHeadset:
        return 1
    case .carAudio, .usb:
        return 2
    case .speakerphone:
        return 3
    case .earpiece:
        return 4
    case .other:
        return 5
    }
}

private func preferredCallAudioRouteDisplay(_ existing: AudioDevice, candidate: AudioDevice) -> AudioDevice {
    let existingRank = callAudioRouteDisplayRank(existing)
    let candidateRank = callAudioRouteDisplayRank(candidate)
    if candidateRank < existingRank { return candidate }
    if candidateRank > existingRank { return existing }
    if candidate.status == .active && existing.status != .active { return candidate }
    return existing
}

private func callAudioRouteDisplayRank(_ device: AudioDevice) -> Int {
    let hasName = !device.displayName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    switch device.kind {
    case .bluetooth(let profile):
        switch profile {
        case .hfp, .ble:
            return hasName ? 0 : 3
        case .unknown:
            return hasName ? 1 : 4
        case .a2dp:
            return hasName ? 2 : 5
        }
    case .speakerphone, .earpiece:
        return 0
    case .wiredHeadset, .carAudio, .usb, .other:
        return hasName ? 0 : 1
    }
}

private func callAudioRouteWithStatus(_ device: AudioDevice, _ status: AudioDeviceStatus) -> AudioDevice {
    AudioDevice(
        id: device.id,
        displayName: device.displayName,
        kind: device.kind,
        direction: device.direction,
        status: status
    )
}

private extension AudioDevice {
    var isCallAudioOutputRoute: Bool {
        direction == .output || direction == .both
    }
}
